lib/spack/spack/repository.py

```diff
@@ -93,6 +101,242 @@ def __getattr__(self, name):
+        cache = {}
+        for pkg_name in os.listdir(self.packages_path):
+            # Skip non-directories in the package root.
+            pkg_dir = join_path(self.packages_path, pkg_name)
+
+            # Warn about invalid names that look like packages.
+            if not valid_module_name(pkg_name):
+                msg = 'Skipping package at {0}. '
+                msg += '"{1]" is not a valid Spack module name.'
+                tty.warn(msg.format(pkg_dir, pkg_name))
+                continue
+
+            # Construct the file name from the directory
+            pkg_file = os.path.join(
+                self.packages_path, pkg_name, package_file_name
+            )
+
+            # Use stat here to avoid lots of calls to the filesystem.
+            try:
+                sinfo = os.stat(pkg_file)
+            except OSError as e:
+                if e.errno == errno.ENOENT:
+                    # No package.py file here.
+                    continue
+                elif e.errno == errno.EACCES:
+                    tty.warn("Can't read package file %s." % pkg_file)
+                    continue
+                raise e
+
+            # If it's not a file, skip it.
+            if stat.S_ISDIR(sinfo.st_mode):
+                continue
+
+            # If it is a file, then save the stats under the
+            # appropriate key
+            cache[pkg_name] = sinfo
+
+        return cache
+
+    def __getitem__(self, item):
+        return self._packages_to_stats[item]
+
+    def __iter__(self):
+        return iter(self._packages_to_stats)
+
+    def __len__(self):
+        return len(self._packages_to_stats)
+
+
+class TagIndex(Mapping):
+    """Maps tags to list of packages."""
+
+    def __init__(self):
+        self._tag_dict = collections.defaultdict(list)
+
+    def to_json(self, stream):
+        json.dump({'tags': self._tag_dict}, stream)
+
+    @staticmethod
+    def from_json(stream):
+        d = json.load(stream)
+
+        r = TagIndex()
+
+        for tag, list in d['tags'].items():
+            r[tag].extend(list)
+
+        return r
+
+    def __getitem__(self, item):
+        return self._tag_dict[item]
+
+    def __iter__(self):
+        return iter(self._tag_dict)
+
+    def __len__(self):
+        return len(self._tag_dict)
+
+    def update_package(self, pkg_name):
+        """Updates a package in the tag index.
+
+        Args:
+            pkg_name (str): name of the package to be removed from the index
+
+        """
+
+        package = spack.repo.get(pkg_name)
+
+        # Remove the package from the list of packages, if present
+        for pkg_list in self._tag_dict.values():
+            if pkg_name in pkg_list:
+                pkg_list.remove(pkg_name)
+
+        # Add it again under the appropriate tags
+        for tag in getattr(package, 'tags', []):
+            self._tag_dict[tag].append(package.name)
+
+
+@llnl.util.lang.memoized
+def make_provider_index_cache(packages_path, namespace):
```

Comment:
@tgamblin This function shares most code with `make_tag_index_cache`. We can refactor them together in a later PR, if you don't think it's a big issue. The idea is that we can take a better look at where `ProviderIndex` is used, and maybe change it to be a `Mapping` too. Then refactor these two functions when the interfaces will be more homogeneous.